ValveAnnulusAnalysis/HeartValveWidgets/ValveSequenceBrowserWidget.py

```diff
@@ -0,0 +1,267 @@
+      self.valveBrowser.valveBrowserNode.SetSelectedItemNumber(heartValveSequenceIndex)
+    else:
+      logging.info(f"Add time point")
+      if not volumeSequenceIndexValue:
+        raise RuntimeError("Failed to add time point, could not get volume sequence")
+      self.valveBrowser.addTimePoint(volumeSequenceIndexValue)
+      heartValveNode = self.valveBrowser.heartValveNode if self.valveBrowser else None
+      self.setHeartValveNode(heartValveNode)
+
+    self.updateGUIFromValveVolumeBrowser()
+
+  def onRemoveTimePointButtonClicked(self):
+    itemIndex, indexValue = self.valveBrowser.getDisplayedHeartValveSequenceIndexAndValue()
+    if indexValue is not None:
+      # TODO: add confirm dialog
+      self.valveBrowser.removeTimePoint(indexValue)
+
+    self.updateGUIFromHeartValveNode()
+
+  def disableGUI(self):
+    self.ui.addTimePointButton.text = "Add volume"
+    self._setCardiacCycleIndex(0)
+    self.ui.setEnabled(False)
+
+  def updateGUIFromValveVolumeBrowser(self, unusedArg1=None, unusedArg2=None, unusedArg3=None):
+    if not self.valveVolumeBrowserNode or self.valveVolumeBrowserNode.GetPlaybackActive():
+      self.disableGUI()
+      return
+
+    seqIndex, seqIndexValue = self.valveBrowser.getDisplayedValveVolumeSequenceIndexAndValue()
+    hasTimepoint = seqIndexValue and self.valveBrowser.heartValveSequenceNode.GetItemNumberFromIndexValue(seqIndexValue) >= 0
+
+    self.ui.setEnabled(self.valveBrowser.valveBrowserNode is not None if self.valveBrowser else False)
+    self.ui.removeTimePointButton.enabled = hasTimepoint
+    self.ui.cardiacCyclePhaseSelector.enabled = hasTimepoint
+
+    t = Template("$action volume (index = $index)")
+    if hasTimepoint:
+      self.ui.addTimePointButton.text = t.substitute(action="Go to", index=seqIndex + 1)
+      heartValveNode = self.valveBrowser.heartValveSequenceNode.GetDataNodeAtValue(seqIndexValue)
+      self._setCardiacCycleIndex(
+        self.ui.cardiacCyclePhaseSelector.findText(heartValveNode.GetAttribute("CardiacCyclePhase"))
+      )
+    else:
+      self.ui.addTimePointButton.text = t.substitute(action="Add", index=seqIndex + 1)
+      self._setCardiacCycleIndex(0)
+
+  def updateGUIFromHeartValveNode(self, unusedArg1=None, unusedArg2=None, unusedArg3=None):
+    self.ui.setEnabled(self.valveBrowser.valveBrowserNode is not None if self.valveBrowser else False)
+
+    self._setCardiacCycleIndex(
+      self.ui.cardiacCyclePhaseSelector.findText(self.valveModel.getCardiacCyclePhase()) if self.valveModel else 0
+    )
+
+    valveVolumeSequenceIndexStr = self.valveModel.getVolumeSequenceIndexAsDisplayedString(
+      self.valveModel.getValveVolumeSequenceIndex()) if self.valveModel else ""
+    self.ui.valveVolumeSequenceIndexValue.setText(valveVolumeSequenceIndexStr)
+
+    valveModelSelected = self.valveModel is not None
+    self.ui.cardiacCyclePhaseSelector.setEnabled(valveModelSelected)
+
+  def _setCardiacCycleIndex(self, index):
```

Comment:
This is an odd-one-out, it would be nice if we could do without it.

Reply:
The following block would be needed 4x in the code. 

```
wasBlocked = self.ui.cardiacCyclePhaseSelector.blockSignals(True)
self.ui.cardiacCyclePhaseSelector.setCurrentIndex({whatever index})
self.ui.cardiacCyclePhaseSelector.blockSignals(wasBlocked)
```

What do you suggest, we do instead of extracting a method? Give it a more specific name? `_setCardiacCycleWithBlockedSignals`?

Reply:
In general, update of a widget from MRML should appear at only a single time in the code. If it only appears at one place and it only 3 lines long then it does not need a dedicated method.

It is not issue to have this method, I just mentioned it because it may be a symptom of some potentially avoidable complexity/redundancy.